File: crates/aicc-hook/src/control.rs
//! The `context` and `canvas` subcommands.
//!
//! Unlike hook mode these are invoked deliberately by an agent, so they are
//! loud: the runtime's prose goes to stdout and any failure goes to stderr with
//! exit code 1.

use std::io::Write;

use serde_json::{json, Map, Value};

use crate::hook::percent_encode_segment;
use crate::http::{self, Request};
use crate::Session;

/// The context-link verbs the runtime exposes.
const CONTEXT_VERBS: [&str; 4] = ["list", "summary", "transcript", "terminal"];

/// `aicc-hook context <verb> [--node <id|title>] [-n N]`
pub fn run_context(args: &[String]) -> i32 {
    let Some(verb) = args.first() else {
        return fail("usage: aicc-hook context <list|summary|transcript|terminal> [--node <id|title>] [-n N]");
    };
    if !CONTEXT_VERBS.contains(&verb.as_str()) {
        return fail(&format!(
            "unknown context verb `{verb}`; expected one of {}",
            CONTEXT_VERBS.join(", ")
        ));
    }

    let mut node: Option<String> = None;
    let mut lines: Option<i64> = None;
    let mut index = 1;
    while index < args.len() {
        let arg = args[index].as_str();
        let (flag, inline) = match arg.split_once('=') {
            Some((flag, value)) if flag.starts_with('-') => (flag, Some(value.to_string())),
            _ => (arg, None),
        };
        match flag {
            "--node" => match take_value(args, &mut index, inline) {
                Some(value) => node = Some(value),
                None => return fail("--node needs a node id or title"),
            },
            "-n" | "--lines" => match take_value(args, &mut index, inline) {
                Some(value) => match value.trim().parse::<i64>() {
                    Ok(parsed) => lines = Some(parsed),
                    Err(_) => return fail(&format!("-n needs a number, got `{value}`")),
                },
                None => return fail("-n needs a number"),
            },
            other => return fail(&format!("unknown option `{other}` for `context {verb}`")),
        }
        index += 1;
    }

    let mut map = Map::new();
    if let Some(node) = node {
        map.insert("node".to_string(), json!(node));
    }
    if let Some(lines) = lines {
        map.insert("n".to_string(), json!(lines));
    }
    request(
        &format!("/context-link/{}", percent_encode_segment(verb)),
        map,
    )
}

/// `aicc-hook canvas <verb> [--flag value | --flag=value | --flag]...`
pub fn run_canvas(args: &[String]) -> i32 {
    let Some(verb) = args.first() else {
        return fail("usage: aicc-hook canvas <verb> [--flag value]...");
    };
    if verb.starts_with('-') {
        return fail(&format!("expected a canvas verb, got `{verb}`"));
    }
    let map = match parse_flags(&args[1..]) {
        Ok(map) => map,
        Err(error) => return fail(&error),
    };
    request(&format!("/control/{}", percent_encode_segment(verb)), map)
}

/// Turns `--flag value`, `--flag=value` and bare `--flag` into an args object.
///
/// A flag repeated more than once collects into an array so verbs such as
/// `link --to a --to b` work without special casing.
pub fn parse_flags(args: &[String]) -> Result<Map<String, Value>, String> {
    let mut map: Map<String, Value> = Map::new();
    let mut index = 0;
    while index < args.len() {
        let arg = args[index].as_str();
        if !arg.starts_with("--") {
            return Err(format!("expected a --flag, got `{arg}`"));
        }
        let (flag, inline) = match arg.split_once('=') {
            Some((flag, value)) => (flag, Some(value.to_string())),
            None => (arg, None),
        };
        let name = flag.trim_start_matches('-');
        if name.is_empty() {
            return Err("`--` is not a flag name".to_string());
        }
        // A flag whose next token is another flag (or nothing) is a boolean.
        let value = match inline {
            Some(value) => Value::String(value),
            None => match args.get(index + 1) {
                Some(next) if !next.starts_with("--") => {
                    index += 1;
                    Value::String(next.clone())
                }
                _ => Value::Bool(true),
            },
        };
        insert_or_append(&mut map, name, value);
        index += 1;
    }
    Ok(map)
}

fn insert_or_append(map: &mut Map<String, Value>, name: &str, value: Value) {
    match map.remove(name) {
        Some(Value::Array(mut existing)) => {
            existing.push(value);
            map.insert(name.to_string(), Value::Array(existing));
        }
        Some(previous) => {
            map.insert(name.to_string(), Value::Array(vec![previous, value]));
        }
        None => {
            map.insert(name.to_string(), value);
        }
    }
}

fn take_value(args: &[String], index: &mut usize, inline: Option<String>) -> Option<String> {
    if let Some(value) = inline {
        return Some(value);
    }
    let next = args.get(*index + 1)?;
    if next.starts_with('-') {
        return None;
    }
    *index += 1;
    Some(next.clone())
}

/// Builds the `{nodeId, args}` body every control route takes.
pub fn control_body(node_id: &str, args: Map<String, Value>) -> Vec<u8> {
    let mut body = Map::new();
    body.insert("nodeId".to_string(), json!(node_id));
    body.insert("args".to_string(), Value::Object(args));
    serde_json::to_vec(&Value::Object(body)).unwrap_or_else(|_| b"{}".to_vec())
}

fn request(path: &str, args: Map<String, Value>) -> i32 {
    let session = match Session::load() {
        Ok(session) => session,
        Err(error) => return fail(&error),
    };
    let body = control_body(&session.node_id, args);
    let request = Request::post_json(path.to_string(), session.headers(), body);
    match http::send(&session.endpoint, &request) {
        Ok(response) if response.is_success() => {
            let text = render(&response);
            if !text.is_empty() {
                let mut stdout = std::io::stdout().lock();
                let _ = stdout.write_all(text.as_bytes());
                if !text.ends_with('\n') {
                    let _ = stdout.write_all(b"\n");
                }
                let _ = stdout.flush();
            }
            0
        }
        Ok(response) => fail(&render_error(&response)),
        Err(error) => fail(&error),
    }
}

/// Renders a successful response: prose is printed as-is, JSON is reduced to
/// its human readable field.
pub fn render(response: &http::Response) -> String {
    if !is_json(response) {
        return response.body.clone();
    }
    let Ok(value) = serde_json::from_str::<Value>(&response.body) else {
        return response.body.clone();
    };
    for key in ["message", "result", "text"] {
        match value.get(key) {
            Some(Value::String(text)) => return text.clone(),
            Some(other) => return other.to_string(),
            None => {}
        }
    }
    response.body.clone()
}

/// Renders a failure response into a single stderr line.
pub fn render_error(response: &http::Response) -> String {
    let fallback = || {
        let body = response.body.trim();
        if body.is_empty() {
            format!("hook endpoint answered {}", response.status)
        } else {
            format!("{} ({})", body, response.status)
        }
    };
    if !is_json(response) {
        return fallback();
    }
    let Ok(value) = serde_json::from_str::<Value>(&response.body) else {
        return fallback();
    };
    for key in ["error", "message"] {
        if let Some(Value::String(text)) = value.get(key) {
            return format!("{text} ({})", response.status);
        }
    }
    fallback()
}

fn is_json(response: &http::Response) -> bool {
    response
        .content_type
        .as_deref()
        .map(|value| value.to_ascii_lowercase().contains("json"))
        .unwrap_or(false)
}

fn fail(message: &str) -> i32 {
    let _ = writeln!(std::io::stderr(), "aicc-hook: {message}");
    1
}

#[cfg(test)]
mod tests {
    use super::*;

    fn flags(input: &[&str]) -> Map<String, Value> {
        let owned: Vec<String> = input.iter().map(|s| s.to_string()).collect();
        parse_flags(&owned).unwrap()
    }

    #[test]
    fn bare_flags_become_true() {
        let map = flags(&["--dry-run"]);
        assert_eq!(map["dry-run"], Value::Bool(true));
    }

    #[test]
    fn both_flag_spellings_agree() {
        assert_eq!(flags(&["--title", "Build"]), flags(&["--title=Build"]));
    }

    #[test]
    fn repeated_flags_collect_into_an_array() {
        let map = flags(&["--after", "a", "--after", "b", "--after=c"]);
        assert_eq!(map["after"], json!(["a", "b", "c"]));
    }

    #[test]
    fn a_flag_followed_by_a_flag_is_boolean() {
        let map = flags(&["--dry-run", "--title", "x"]);
        assert_eq!(map["dry-run"], Value::Bool(true));
        assert_eq!(map["title"], json!("x"));
    }

    #[test]
    fn values_may_look_like_short_options() {
        // Only `--` prefixed tokens end a value, so a title of "-5" survives.
        let map = flags(&["--title=-5"]);
        assert_eq!(map["title"], json!("-5"));
    }

    #[test]
    fn positional_arguments_are_rejected() {
        let owned = vec!["oops".to_string()];
        assert!(parse_flags(&owned).is_err());
    }

    #[test]
    fn control_body_shape() {
        // `serde_json` maps are ordered, so the bytes are reproducible; the
        // runtime only cares about the field names.
        let body = control_body("n1", flags(&["--dry-run"]));
        assert_eq!(
            String::from_utf8(body).unwrap(),
            r#"{"args":{"dry-run":true},"nodeId":"n1"}"#
        );
    }

    #[test]
    fn control_body_keeps_an_empty_args_object() {
        let body = control_body("n1", Map::new());
        assert_eq!(
            String::from_utf8(body).unwrap(),
            r#"{"args":{},"nodeId":"n1"}"#
        );
    }

    #[test]
    fn json_responses_are_reduced_to_their_message() {
        let response = http::Response {
            status: 200,
            content_type: Some("application/json".to_string()),
            body: r#"{"ok":true,"message":"opened 2 nodes"}"#.to_string(),
        };
        assert_eq!(render(&response), "opened 2 nodes");
    }

    #[test]
    fn plain_text_responses_pass_through() {
        let response = http::Response {
            status: 200,
            content_type: Some("text/plain; charset=utf-8".to_string()),
            body: "linked nodes:\n- api (n2)\n".to_string(),
        };
        assert_eq!(render(&response), "linked nodes:\n- api (n2)\n");
    }

    #[test]
    fn errors_carry_the_status() {
        let json = http::Response {
            status: 403,
            content_type: Some("application/json".to_string()),
            body: r#"{"error":"node is not linked"}"#.to_string(),
        };
        assert_eq!(render_error(&json), "node is not linked (403)");

        let plain = http::Response {
            status: 500,
            content_type: Some("text/plain".to_string()),
            body: "boom".to_string(),
        };
        assert_eq!(render_error(&plain), "boom (500)");

        let empty = http::Response {
            status: 502,
            content_type: None,
            body: String::new(),
        };
        assert_eq!(render_error(&empty), "hook endpoint answered 502");
    }
}

File: apps/desktop/src/cli/armadra-hook/control.ts
/**
 * The `context`, `canvas` and `browser` subcommands.
 *
 * Unlike hook mode these are invoked deliberately by an agent, so they are
 * loud: the runtime's prose goes to stdout and any failure goes to stderr with
 * exit code 1.
 *
 * The canvas verb itself is never validated here. The runtime owns that
 * registry — `canvas help` is served from it — so a client-side copy would be
 * a second source of truth that drifts the first time a verb is added.
 */

import { envVar } from "./endpoint.js";
import { percentEncodeSegment } from "./hook.js";
import { canonicalJsonBytes, tryParseJson } from "./json.js";
import type { JsonValue } from "./json.js";
import { isSuccess, postJsonRequest, totalTimeoutMs } from "./http.js";
import type { HookResponse } from "./http.js";
import { headersFor, loadSession, send } from "./session.js";
import { FILE_SUFFIX, STDIN_VALUE, TextReader } from "./text-input.js";
import type { TextSources } from "./text-input.js";
import { BROWSER_VERBS, CONTEXT_VERBS } from "./usage.js";

/** The `{flag: value}` object a control route carries as `args`. */
export type Args = Record<string, JsonValue>;

/** `armadra-hook context <verb> [--node <id|title>] [-n N]` */
export async function runContext(args: string[]): Promise<number> {
  const verb = args[0];
  if (verb === undefined) {
    return fail(
      "usage: armadra-hook context <list|summary|transcript|terminal> [--node <id|title>] [-n N] [--since] [--full --max-kb N]",
    );
  }
  if (!(CONTEXT_VERBS as readonly string[]).includes(verb)) {
    return fail(
      `unknown context verb \`${verb}\`; expected one of ${CONTEXT_VERBS.join(", ")}`,
    );
  }

  let node: string | undefined;
  let lines: number | undefined;
  // 阶段 C+ 的三个读取旋钮（设计 agent-delivery.md §13）。
  let since = false;
  let full = false;
  let maxKb: number | undefined;
  let index = 1;
  while (index < args.length) {
    const arg = args[index]!;
    const separator = arg.indexOf("=");
    const inline =
      separator > 0 && arg.startsWith("-")
        ? arg.slice(separator + 1)
        : undefined;
    const flag = inline === undefined ? arg : arg.slice(0, separator);
    if (flag === "--node") {
      const cursor = { index };
      const value = takeValue(args, cursor, inline);
      index = cursor.index;
      if (value === undefined) return fail("--node needs a node id or title");
      node = value;
    } else if (flag === "-n" || flag === "--lines") {
      const cursor = { index };
      const value = takeValue(args, cursor, inline);
      index = cursor.index;
      if (value === undefined) return fail("-n needs a number");
      const parsed = parseInteger(value);
      if (parsed === undefined)
        return fail(`-n needs a number, got \`${value}\``);
      lines = parsed;
    } else if (flag === "--since") {
      since = true;
    } else if (flag === "--full") {
      full = true;
    } else if (flag === "--max-kb") {
      const cursor = { index };
      const value = takeValue(args, cursor, inline);
      index = cursor.index;
      if (value === undefined) return fail("--max-kb needs a number");
      const parsed = parseInteger(value);
      if (parsed === undefined)
        return fail(`--max-kb needs a number, got \`${value}\``);
      maxKb = parsed;
    } else {
      return fail(`unknown option \`${flag}\` for \`context ${verb}\``);
    }
    index += 1;
  }

  const map: Args = {};
  if (node !== undefined) map["node"] = node;
  if (lines !== undefined) map["n"] = lines;
  if (since) map["since"] = true;
  if (full) map["full"] = true;
  if (maxKb !== undefined) map["max-kb"] = maxKb;
  return request(`/context-link/${percentEncodeSegment(verb)}`, map);
}

/** `armadra-hook canvas <verb> [--flag value | --flag=value | --flag]...` */
export async function runCanvas(args: string[]): Promise<number> {
  const verb = args[0];
  if (verb === undefined)
    return fail("usage: armadra-hook canvas <verb> [--flag value]...");
  if (verb.startsWith("-"))
    return fail(`expected a canvas verb, got \`${verb}\``);
  const parsed = parseFlags(args.slice(1));
  if ("error" in parsed) return fail(parsed.error);
  const map = parsed.ok;
  if (verb === "handoff-read" || verb === "ack") {
    delete map["sessionId"];
    delete map["generation"];
    const session = envVar("ARMADRA_SESSION_ID");
    const generation = parseUnsigned(envVar("ARMADRA_SESSION_GENERATION"));
    if (session !== undefined && generation !== undefined) {
      map["sessionId"] = session;
      map["generation"] = generation;
    } else if (verb === "handoff-read") {
      return fail(
        "A current terminal session binding is required; restart an older terminal.",
      );
    }
  }
  return request(`/control/${percentEncodeSegment(verb)}`, map);
}

/**
 * `armadra-hook browser <verb> [--flag value]...`
 *
 * Drives a browser node this node is linked to on the canvas. The same session
 * a person is looking at — there is no separate agent browser.
 */
/**
 * How long one browser verb may take, per endpoint candidate.
 *
 * NOT the hook's 1.5 s. That budget is for a hook event, which must never
 * hold up the CLI that fired it; a browser verb is deliberately long — `wait`
 * waits up to 30 s, a navigation waits for the page, the shell gives a verb
 * 45 s — and a request that runs out of budget here does not fail cleanly: the
 * client moves on to the next candidate and sends the verb AGAIN, which for a
 * click means clicking twice. So the budget covers the longest verb the
 * runtime allows plus its own timeout. `ARMADRA_HOOK_TIMEOUT_MS` still
 * overrides it when set.
 */
export const BROWSER_TIMEOUT_MS = 70_000;

export function browserTimeoutMs(): number {
  return envVar("ARMADRA_HOOK_TIMEOUT_MS") === undefined
    ? BROWSER_TIMEOUT_MS
    : totalTimeoutMs();
}

export async function runBrowser(args: string[]): Promise<number> {
  const verb = args[0];
  if (verb === undefined) {
    return fail(
      `usage: armadra-hook browser <${BROWSER_VERBS.join("|")}> [--flag value]...`,
    );
  }
  if (!BROWSER_VERBS.includes(verb)) {
    return fail(
      `unknown browser verb \`${verb}\`; expected one of ${BROWSER_VERBS.join(", ")}`,
    );
  }
  // `-n N` is the one short flag the browser verbs take (`read --limit`); the
  // rest of the parser only knows `--flags`.
  const parsed = parseFlags(
    args.slice(1).map((arg) => (arg === "-n" ? "--limit" : arg)),
  );
  if ("error" in parsed) return fail(parsed.error);
  return request(
    `/browser/${percentEncodeSegment(verb)}`,
    parsed.ok,
    browserTimeoutMs(),
  );
}

/**
 * Turns `--flag value`, `--flag=value` and bare `--flag` into an args object.
 *
 * A flag repeated more than once collects into an array so verbs such as
 * `link --to a --to b` work without special casing.
 *
 * `--flag -` takes the value from stdin and `--flag-file PATH` from a file
 * (`text-input.ts`): arbitrary text — a message body, a task — does not have
 * to survive the shell's quoting, and on Windows `cmd.exe`'s.
 */
export function parseFlags(
  args: string[],
  sources?: TextSources,
): { ok: Args } | { error: string } {
  const reader = new TextReader(sources);
  const map: Args = {};
  let index = 0;
  while (index < args.length) {
    const arg = args[index]!;
    if (!arg.startsWith("--"))
      return { error: `expected a --flag, got \`${arg}\`` };
    const separator = arg.indexOf("=");
    const inline = separator >= 0 ? arg.slice(separator + 1) : undefined;
    const flag = separator >= 0 ? arg.slice(0, separator) : arg;
    const name = flag.replace(/^-+/, "");
    if (name === "") return { error: "`--` is not a flag name" };
    // A flag whose next token is another flag (or nothing) is a boolean.
    let value: JsonValue;
    if (inline !== undefined) {
      value = inline;
    } else {
      const next = args[index + 1];
      if (next !== undefined && !next.startsWith("--")) {
        index += 1;
        value = next;
      } else {
        value = true;
      }
    }
    let target = name;
    if (name.length > FILE_SUFFIX.length && name.endsWith(FILE_SUFFIX)) {
      target = name.slice(0, -FILE_SUFFIX.length);
      if (typeof value !== "string" || value === "")
        return { error: `--${name} needs a path` };
      const read = reader.file(target, value);
      if ("error" in read) return read;
      value = read.ok;
    } else if (value === STDIN_VALUE) {
      const read = reader.stdin(name);
      if ("error" in read) return read;
      value = read.ok;
    }
    insertOrAppend(map, target, value);
    index += 1;
  }
  return { ok: map };
}

function insertOrAppend(map: Args, name: string, value: JsonValue): void {
  const previous = map[name];
  if (previous === undefined) {
    map[name] = value;
  } else if (Array.isArray(previous)) {
    map[name] = [...previous, value];
  } else {
    map[name] = [previous, value];
  }
}

function takeValue(
  args: string[],
  cursor: { index: number },
  inline: string | undefined,
): string | undefined {
  if (inline !== undefined) return inline;
  const next = args[cursor.index + 1];
  if (next === undefined || next.startsWith("-")) return undefined;
  cursor.index += 1;
  return next;
}

/** `str::parse::<i64>` — a sign, digits, nothing else. */
function parseInteger(value: string): number | undefined {
  const text = value.trim();
  if (!/^[+-]?\d+$/.test(text)) return undefined;
  const parsed = Number(text);
  return Number.isSafeInteger(parsed) ? parsed : undefined;
}

/** `str::parse::<u64>` — digits only, no sign. */
function parseUnsigned(value: string | undefined): number | undefined {
  if (value === undefined) return undefined;
  if (!/^\d+$/.test(value)) return undefined;
  const parsed = Number(value);
  return Number.isSafeInteger(parsed) ? parsed : undefined;
}

/** Builds the `{nodeId, args}` body every control route takes. */
export function controlBody(nodeId: string, args: Args): Buffer {
  return canonicalJsonBytes({ nodeId, args });
}

async function request(
  path: string,
  args: Args,
  total?: number,
): Promise<number> {
  const loaded = loadSession();
  if ("error" in loaded) return fail(loaded.error);
  const session = loaded.ok;
  const body = controlBody(session.nodeId, args);
  const outcome = await send(
    session,
    (current, candidate) =>
      postJsonRequest(path, headersFor(current, candidate), body),
    total,
  );
  if ("error" in outcome) return fail(outcome.error);
  if (!isSuccess(outcome.ok)) return fail(renderError(outcome.ok));
  const text = render(outcome.ok);
  if (text !== "")
    process.stdout.write(text.endsWith("\n") ? text : `${text}\n`);
  return 0;
}

/**
 * Renders a successful response: prose is printed as-is, JSON is reduced to
 * its human readable field.
 */
export function render(response: HookResponse): string {
  if (!isJson(response)) return response.body;
  const value = tryParseJson(response.body);
  if (
    value === undefined ||
    value === null ||
    typeof value !== "object" ||
    Array.isArray(value)
  ) {
    return response.body;
  }
  const object = value as Record<string, JsonValue>;
  if ("protocol" in object || "outcome" in object) return response.body;
  for (const key of ["message", "result", "text"]) {
    if (!(key in object)) continue;
    const field = object[key];
    return typeof field === "string" ? field : jsonToString(field);
  }
  return response.body;
}

/** `serde_json::Value::to_string` for the non-string branch of `render`. */
function jsonToString(value: JsonValue | undefined): string {
  if (value === undefined) return "null";
  // Rust prints the value with its own key order (sorted), which is what
  // `canonicalJson` produces.
  return canonicalJsonBytes(value).toString("utf8");
}

/** Renders a failure response into a single stderr line. */
export function renderError(response: HookResponse): string {
  const fallback = (): string => {
    const body = response.body.trim();
    return body === ""
      ? `hook endpoint answered ${response.status}`
      : `${body} (${response.status})`;
  };
  if (!isJson(response)) return fallback();
  const value = tryParseJson(response.body);
  if (
    value === undefined ||
    value === null ||
    typeof value !== "object" ||
    Array.isArray(value)
  ) {
    return fallback();
  }
  const object = value as Record<string, JsonValue>;
  for (const key of ["error", "message"]) {
    const field = object[key];
    if (typeof field === "string") return `${field} (${response.status})`;
  }
  return fallback();
}

function isJson(response: HookResponse): boolean {
  return (response.contentType ?? "").toLowerCase().includes("json");
}

export function fail(message: string): number {
  process.stderr.write(`armadra-hook: ${message}\n`);
  return 1;
}
